File: samples/python/tutorial_code/features2D/Homography/perspective_correction.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

# Python 2/3 compatibility
from __future__ import print_function

import numpy as np
import cv2 as cv
import sys


def randomColor():
    color = np.random.randint(0, 255,(1, 3))
    return color[0].tolist()

def  perspectiveCorrection(img1Path, img2Path ,patternSize ):
    img1 = cv.imread(cv.samples.findFile(img1Path))
    img2 = cv.imread(cv.samples.findFile(img2Path))

    # [find-corners]
    ret1, corners1 = cv.findChessboardCorners(img1, patternSize)
    ret2, corners2 = cv.findChessboardCorners(img2, patternSize)
    # [find-corners]

    if not ret1 or not ret2:
        print("Error, cannot find the chessboard corners in both images.")
        sys.exit(-1)

    # [estimate-homography]
    H, _ = cv.findHomography(corners1, corners2)
    print(H)
    # [estimate-homography]

    # [warp-chessboard]
    img1_warp = cv.warpPerspective(img1, H, (img1.shape[1], img1.shape[0]))
    # [warp-chessboard]

    img_draw_warp = cv.hconcat([img2, img1_warp])
    cv.imshow("Desired chessboard view / Warped source chessboard view", img_draw_warp )

    corners1 = corners1.tolist()
    corners1 = [a[0] for a in corners1]

    # [compute-transformed-corners]
    img_draw_matches = cv.hconcat([img1, img2])
    for i in range(len(corners1)):
        pt1 = np.array([corners1[i][0], corners1[i][1], 1])
        pt1 = pt1.reshape(3, 1)
        pt2 = np.dot(H, pt1)
        pt2 = pt2/pt2[2]
        end = (int(img1.shape[1] + pt2[0]), int(pt2[1]))
        cv.line(img_draw_matches, tuple([int(j) for j in corners1[i]]), end, randomColor(), 2)

    cv.imshow("Draw matches", img_draw_matches)
    cv.waitKey(0)
    # [compute-transformed-corners]

def main():
    import argparse
    parser = argparse.ArgumentParser()
    parser.add_argument('-I1', "--image1", help="Path to the first image", default="left02.jpg")
    parser.add_argument('-I2', "--image2", help="Path to the second image", default="left01.jpg")
    parser.add_argument('-H', "--height", help="Height of pattern size", default=6)
    parser.add_argument('-W', "--width", help="Width of pattern size", default=9)
    args = parser.parse_args()

    img1Path = args.image1
    img2Path = args.image2
    h = args.height
    w = args.width
    perspectiveCorrection(img1Path, img2Path, (w, h))

if __name__ == "__main__":
    main()

File: samples/cpp/tutorial_code/features2D/Homography/perspective_correction.cpp
#include <iostream>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/calib3d.hpp>
#include <opencv2/highgui.hpp>

using namespace std;
using namespace cv;

namespace
{
enum Pattern { CHESSBOARD, CIRCLES_GRID, ASYMMETRIC_CIRCLES_GRID };

Scalar randomColor( RNG& rng )
{
  int icolor = (unsigned int) rng;
  return Scalar( icolor & 255, (icolor >> 8) & 255, (icolor >> 16) & 255 );
}

void perspectiveCorrection(const string &img1Path, const string &img2Path, const Size &patternSize, RNG &rng)
{
    Mat img1 = imread( samples::findFile(img1Path) );
    Mat img2 = imread( samples::findFile(img2Path) );

    //! [find-corners]
    vector<Point2f> corners1, corners2;
    bool found1 = findChessboardCorners(img1, patternSize, corners1);
    bool found2 = findChessboardCorners(img2, patternSize, corners2);
    //! [find-corners]

    if (!found1 || !found2)
    {
        cout << "Error, cannot find the chessboard corners in both images." << endl;
        return;
    }

    //! [estimate-homography]
    Mat H = findHomography(corners1, corners2);
    cout << "H:\n" << H << endl;
    //! [estimate-homography]

    //! [warp-chessboard]
    Mat img1_warp;
    warpPerspective(img1, img1_warp, H, img1.size());
    //! [warp-chessboard]

    Mat img_draw_warp;
    hconcat(img2, img1_warp, img_draw_warp);
    imshow("Desired chessboard view / Warped source chessboard view", img_draw_warp);

    //! [compute-transformed-corners]
    Mat img_draw_matches;
    hconcat(img1, img2, img_draw_matches);
    for (size_t i = 0; i < corners1.size(); i++)
    {
        Mat pt1 = (Mat_<double>(3,1) << corners1[i].x, corners1[i].y, 1);
        Mat pt2 = H * pt1;
        pt2 /= pt2.at<double>(2);

        Point end( (int) (img1.cols + pt2.at<double>(0)), (int) pt2.at<double>(1) );
        line(img_draw_matches, corners1[i], end, randomColor(rng), 2);
    }

    imshow("Draw matches", img_draw_matches);
    waitKey();
    //! [compute-transformed-corners]
}

const char* params
    = "{ help h         |       | print usage }"
      "{ image1         | left02.jpg | path to the source chessboard image }"
      "{ image2         | left01.jpg | path to the desired chessboard image }"
      "{ width bw       | 9     | chessboard width }"
      "{ height bh      | 6     | chessboard height }";
}

int main(int argc, char *argv[])
{
    cv::RNG rng( 0xFFFFFFFF );
    CommandLineParser parser(argc, argv, params);

    if (parser.has("help"))
    {
        parser.about("Code for homography tutorial.\n"
            "Example 2: perspective correction.\n");
        parser.printMessage();
        return 0;
    }

    Size patternSize(parser.get<int>("width"), parser.get<int>("height"));
    perspectiveCorrection(parser.get<String>("image1"),
                          parser.get<String>("image2"),
                          patternSize, rng);

    return 0;
}
